onadata/libs/utils/logger_tools.py

```diff
@@ -127,29 +127,54 @@ def get_uuid_from_submission(xml):
     return len(split_xml) > 1 and split_xml[1] or None
 
 
-def get_xform_from_submission(xml, username, uuid=None):
-    # check alternative form submission ids
-    uuid = uuid or get_uuid_from_submission(xml)
-
-    if not username and not uuid:
+def get_xform_from_submission(
+        xml, request, uuid=None, owner_username=None):
+    """Gets the submissions target XForm.
+
+    Retrieves the target XForm by either utilizing the `uuid` param
+    or the `uuid` retrievable from the `xml` or the `id_string`
+    retrievable from the XML. Only returns form if `request_user` has
+    permission to submit.
+
+    :param (str) xml: The submission in XML form
+    :param (django.http.request) request: Request object
+    :param (str) uuid: The target XForms universally unique identifier.
+    Default: None
+    :param (str) owner_username: The owner of the target XForm. Default: None
+    """
+    if not uuid and not (owner_username or request.user):
         raise InstanceInvalidUserError()
 
+    username = None
+    if owner_username:
+        username = owner_username
+    elif request.user.is_authenticated:
+        username = request.user.username
+
     if uuid:
         # try find the form by its uuid which is the ideal condition
         if XForm.objects.filter(
                 uuid=uuid, deleted_at__isnull=True).count() > 0:
             xform = XForm.objects.get(uuid=uuid, deleted_at__isnull=True)
-
-            return xform
+            try:
+                # Verify request user has permission to make submissions to the
+                # XForm
+                check_submission_permissions(request, xform)
+                return xform
+            except PermissionDenied as e:
+                # Check if the owner_username is equal to the XForm owner
+                # Assumption: If the owner_username is equal to the XForm
+                # owner we've retrieved the correct form.
+                if owner_username and xform.user.username == owner_username:
+                    raise e
```

Comment:
I'm confused a bit here. Why are we raising an error if the owner username is the same as the xform user?

Reply:
We raise the permission error since we've technically retrieved the correct form the user was trying to make a submission too and they don't have the permission required to make the submission